extensions/opc/src/fetchopc.cpp

```diff
@@ -0,0 +1,239 @@
+#include "core/Processor.h"
+#include "core/ProcessSession.h"
+#include "core/Core.h"
+#include "core/Property.h"
+#include "core/Resource.h"
+#include "controllers/SSLContextService.h"
+#include "core/logging/LoggerConfiguration.h"
+#include "utils/Id.h"
+#include "utils/StringUtils.h"
+
+namespace org {
+namespace apache {
+namespace nifi {
+namespace minifi {
+namespace processors {
+  core::Property FetchOPCProcessor::NodeID(
+      core::PropertyBuilder::createProperty("Node ID")
+      ->withDescription("Specifies the ID of the root node to traverse")
+      ->isRequired(true)->build());
+
+
+  core::Property FetchOPCProcessor::NodeIDType(
+      core::PropertyBuilder::createProperty("Node ID type")
+      ->withDescription("Specifies the type of the provided node ID")
+      ->isRequired(true)
+      ->withAllowableValues<std::string>({"Path", "Int", "String"})->build());
+
+  core::Property FetchOPCProcessor::NameSpaceIndex(
+      core::PropertyBuilder::createProperty("Namespace index")
+      ->withDescription("The index of the namespace. Used only if node ID type is not path.")
+      ->withDefaultValue<int32_t>(0)->build());
+
+  core::Relationship FetchOPCProcessor::Success("success", "Successfully retrieved OPC-UA nodes");
+  core::Relationship FetchOPCProcessor::Failure("failure", "Retrieved OPC-UA nodes where value cannot be extracted (only if enabled)");
+
+
+  void FetchOPCProcessor::initialize() {
+    //BaseOPCProcessor::initialize();
+
+    // Set the supported properties
+    std::set<core::Property> fetchOPCProperties = {OPCServerEndPoint, NodeID, NodeIDType, NameSpaceIndex};
+    std::set<core::Property> baseOPCProperties = BaseOPCProcessor::getSupportedProperties();
+    fetchOPCProperties.insert(baseOPCProperties.begin(), baseOPCProperties.end());
+    setSupportedProperties(fetchOPCProperties);
+
+    // Set the supported relationships
+    setSupportedRelationships({Success, Failure});
+  }
+
+  void FetchOPCProcessor::onSchedule(const std::shared_ptr<core::ProcessContext> &context, const std::shared_ptr<core::ProcessSessionFactory> &factory) {
+    logger_->log_trace("FetchOPCProcessor::onSchedule");
+
+    translatedNodeIDs_.clear();  // Path might has changed during restart
+
+    BaseOPCProcessor::onSchedule(context, factory);
+
+    if(!configOK_) {
+      return;
+    }
+
+    configOK_ = false;
+
+    std::string value;
+    context->getProperty(NodeID.getName(), nodeID_);
+    context->getProperty(NodeIDType.getName(), value);
+
+    if (value == "String") {
+      idType_ = opc::OPCNodeIDType::String;
+    } else if (value == "Int") {
+      idType_ = opc::OPCNodeIDType::Int;
+    } else if (value == "Path") {
+      idType_ = opc::OPCNodeIDType::Path;
+    } else {
+      // Where have our validators gone?
+      logger_->log_error("%s is not a valid node ID type!", value.c_str());
+    }
+
+    if(idType_ == opc::OPCNodeIDType::Int) {
+      try {
+        int t = std::stoi(nodeID_);
```

Comment:
This doesn't perfectly validate whether nodeID_ is an integer, as std::stoi stops at the first character it couldn't convert, so to actually make sure, you have to use its `std::size_t* pos` out parameter to find out whether it consumed the entire string.
Not a big deal, though, I think it is a fair check as it is.

Reply:
That's a fair point, but I would prefer to leave that as a follow-up in a generic way: create a wrapper around stoi in our utils and use that instead of stoi in such places. There are many I guess.